File: scripts/lib/conversation.py
import os
import json
import sqlite3
from typing import Optional, List, Dict, Any, Generator

from .proto_decoder import extract_step_payload

class ConversationDataAccessLayer:
    """
    CDAL: Conversation Data Access Layer
    A centralized facade for accessing Antigravity conversation history and metadata.
    Currently implements a hybrid CQRS transition strategy:
    - 0-latency operations (like status=5 check) use SQLite directly.
    - All JSONL reads have been physically replaced by zero-latency SQLite blob decoding.
    """
    
    def __init__(self, conv_id: str):
        self.conv_id = conv_id
        home_dir = os.environ.get("HOME", os.path.expanduser("~"))
        self.brain_dir = os.path.join(home_dir, ".gemini", "antigravity", "brain")
        
        # [SSOT] Zero-latency synchronous database
        self.db_path = os.path.join(home_dir, ".gemini", "antigravity", "conversations", f"{conv_id}.db")

    # ---------------------------------------------------------
    # 1. Database Metadata
    # ---------------------------------------------------------
    
    def get_compaction_watermark(self) -> int:
        """
        [SSOT] Probe the latest compaction boundary.
        Returns the maximum idx of steps that have been compacted (status=5).
        Returns -1 if no DB exists or no compacted steps exist.
        """
        if not os.path.exists(self.db_path):
            return -1
        try:
            from contextlib import closing
            with closing(sqlite3.connect(self.db_path, timeout=15)) as conn:
                with conn:
                    cursor = conn.cursor()
                    cursor.execute("SELECT MAX(idx) FROM steps WHERE status = 5;")
                    row = cursor.fetchone()
                    return row[0] if row and row[0] is not None else -1
        except Exception:
            return -1

    def get_max_step_index(self) -> int:
        """
        Returns the max step index in the DB.
        """
        if not os.path.exists(self.db_path):
            return 0
        try:
            from contextlib import closing
            with closing(sqlite3.connect(self.db_path, timeout=15)) as conn:
                cursor = conn.cursor()
                cursor.execute("SELECT MAX(idx) FROM steps")
                row = cursor.fetchone()
                return row[0] if row and row[0] is not None else 0
        except Exception:
            return 0

    def get_db_mtime(self) -> float:
        """
        Returns the modification time of the SQLite database.
        Used for high-frequency zombie/idle detection without Ghost Delay.
        """
        if os.path.exists(self.db_path):
            return os.path.getmtime(self.db_path)
        return 0.0

    # ---------------------------------------------------------
    # 2. Native SQLite Payload Extraction
    # ---------------------------------------------------------
    
    def stream_steps_reverse(self, limit: int = 1000) -> Generator[Dict[str, Any], None, None]:
        """
        Yields parsed step dicts from the SQLite steps table in reverse order (DESC).
        Replaces legacy JSONL stream_transcript_reverse.
        """
        if not os.path.exists(self.db_path):
            return
            
        try:
            from contextlib import closing
            with closing(sqlite3.connect(self.db_path, timeout=15)) as conn:
                cursor = conn.cursor()
                cursor.execute("SELECT idx, step_payload FROM steps ORDER BY idx DESC LIMIT ?", (limit,))
                for row in cursor:
                    idx, blob = row
                    entry = extract_step_payload(blob)
                    entry['step_index'] = idx
                    # Some clients check 'step_type' from DB rather than parsed 'type'
                    yield entry
        except Exception:
            return

    def stream_steps_forward(self, start_idx: int = 0) -> Generator[Dict[str, Any], None, None]:
        """
        Yields parsed step dicts in chronological order.
        Replaces legacy JSONL stream_history.
        """
        if not os.path.exists(self.db_path):
            return
            
        try:
            from contextlib import closing
            with closing(sqlite3.connect(self.db_path, timeout=15)) as conn:
                cursor = conn.cursor()
                cursor.execute("SELECT idx, step_payload FROM steps WHERE idx >= ? ORDER BY idx ASC", (start_idx,))
                for row in cursor:
                    idx, blob = row
                    entry = extract_step_payload(blob)
                    entry['step_index'] = idx
                    yield entry
        except Exception:
            return

    def get_latest_user_message(self) -> Optional[str]:
        """
        Extracts the most recent USER_INPUT message content.
        """
        for step in self.stream_steps_reverse(limit=50):
            if step.get("type") == "USER_INPUT":
                return step.get("content", "")
        return None

    def get_latest_planner_response(self) -> Optional[str]:
        """
        Extracts the most recent PLANNER_RESPONSE text.
        """
        for step in self.stream_steps_reverse(limit=50):
            if step.get("type") == "PLANNER_RESPONSE":
                return step.get("content", "")
        return None

    def get_current_turn_idx(self) -> int:
        """
        [SSOT] Returns the idx of the latest USER_INPUT step, representing the current Turn ID.
        """
        for step in self.stream_steps_reverse(limit=1000):
            if step.get("type") == "USER_INPUT":
                return step.get("step_index", 0)
        return 0

    def get_user_input_count(self) -> int:
        """
        Returns the total number of USER_INPUT steps in the conversation.
        """
        count = 0
        for step in self.stream_steps_forward():
            if step.get("type") == "USER_INPUT":
                count += 1
        return count

File: packages/adapter-antigravity/src/bridge/conversation.ts
import * as crypto from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";
import Database from "better-sqlite3";
import { extractStepPayload, parseProtobuf } from "./step-codec";
import { getBrainDir, getConversationsDir } from "./paths";
const PB_KEY = Buffer.from("safeCodeiumworldKeYsecretBalloon", "utf8");

function decryptPb(data: Buffer): Buffer {
	const iv = data.subarray(0, 12);
	const tag = data.subarray(data.length - 16);
	const ciphertext = data.subarray(12, data.length - 16);
	const decipher = crypto.createDecipheriv("aes-256-gcm", PB_KEY, iv);
	decipher.setAuthTag(tag);
	return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
}

export class ConversationDataAccessLayer {
	convId: string;
	dbPath: string;
	pbPath: string;
	private _pbSteps: Record<string, any>[] | null = null;

	constructor(convId: string) {
		this.convId = convId;
		this.dbPath = path.join(getConversationsDir(), `${convId}.db`);
		this.pbPath = path.join(getConversationsDir(), `${convId}.pb`);
	}

	private _loadPbSteps(): Record<string, any>[] {
		if (this._pbSteps !== null) return this._pbSteps;
		const root = this.getPbRoot();
		if (!root) {
			this._pbSteps = [];
			return [];
		}
		const blobs = root[2] ?? [];
		this._pbSteps = blobs.map((b: Buffer, i: number) => {
			const entry = extractStepPayload(b);
			entry["step_index"] = i + 1;
			return entry;
		});
		return this._pbSteps;
	}

	public exists(): boolean {
		return this._dbHasContent() || this.hasPb();
	}

	private _dbHasContent(): boolean {
		if (!fs.existsSync(this.dbPath)) return false;
		try {
			return fs.statSync(this.dbPath).size > 0;
		} catch {
			return false;
		}
	}

	// ---------------------------------------------------------
	// 1. Database Metadata
	// ---------------------------------------------------------

	getCompactionWatermark(): number {
		if (!fs.existsSync(this.dbPath)) {
			return -1;
		}
		try {
			const db = new Database(this.dbPath, { timeout: 15000 });
			try {
				const row = db
					.prepare("SELECT MAX(idx) FROM steps WHERE status = 5;")
					.get() as { "MAX(idx)": number | null } | undefined;
				return row && row["MAX(idx)"] !== null ? row["MAX(idx)"] : -1;
			} finally {
				db.close();
			}
		} catch {
			return -1;
		}
	}

	getMaxStepIndex(): number {
		if (this._dbHasContent()) {
			try {
				const db = new Database(this.dbPath, { timeout: 15000 });
				try {
					const row = db.prepare("SELECT MAX(idx) FROM steps").get() as
						| { "MAX(idx)": number | null }
						| undefined;
					return row && row["MAX(idx)"] !== null ? row["MAX(idx)"] : 0;
				} finally {
					db.close();
				}
			} catch {
				// fall through to PB
			}
		}
		// PB fallback
		return this._loadPbSteps().length;
	}

	getDbMtime(): number {
		if (fs.existsSync(this.dbPath)) {
			return fs.statSync(this.dbPath).mtimeMs / 1000;
		}
		if (fs.existsSync(this.pbPath)) {
			return fs.statSync(this.pbPath).mtimeMs / 1000;
		}
		return 0;
	}

	// ---------------------------------------------------------
	// 2. Native SQLite Payload Extraction
	// ---------------------------------------------------------

	*streamStepsReverse(limit: number = 1000): Generator<Record<string, any>> {
		if (this._dbHasContent()) {
			try {
				const db = new Database(this.dbPath, { timeout: 15000 });
				try {
					const rows = db
						.prepare(
							"SELECT idx, step_payload FROM steps ORDER BY idx DESC LIMIT ?",
						)
						.all(limit) as Array<{ idx: number; step_payload: Buffer }>;
					for (const row of rows) {
						const entry = extractStepPayload(row.step_payload);
						entry["step_index"] = row.idx;
						yield entry;
					}
					return;
				} finally {
					db.close();
				}
			} catch {
				// fall through to PB
			}
		}
		// PB fallback
		const steps = this._loadPbSteps();
		let yielded = 0;
		for (let i = steps.length - 1; i >= 0 && yielded < limit; i--) {
			yield steps[i];
			yielded++;
		}
	}

	*streamStepsForward(startIdx: number = 0): Generator<Record<string, any>> {
		if (this._dbHasContent()) {
			try {
				const db = new Database(this.dbPath, { timeout: 15000 });
				try {
					const rows = db
						.prepare(
							"SELECT idx, step_payload FROM steps WHERE idx >= ? ORDER BY idx ASC",
						)
						.all(startIdx) as Array<{ idx: number; step_payload: Buffer }>;
					for (const row of rows) {
						const entry = extractStepPayload(row.step_payload);
						entry["step_index"] = row.idx;
						yield entry;
					}
					return;
				} finally {
					db.close();
				}
			} catch {
				// fall through to PB
			}
		}
		// PB fallback
		for (const s of this._loadPbSteps()) {
			if (s["step_index"] >= startIdx) {
				yield s;
			}
		}
	}

	getLatestUserMessage(): string | null {
		for (const step of this.streamStepsReverse(50)) {
			if (step["type"] === "USER_INPUT") {
				return step["content"] ?? "";
			}
		}
		return null;
	}

	getLatestPlannerResponse(): string | null {
		for (const step of this.streamStepsReverse(50)) {
			if (step["type"] === "PLANNER_RESPONSE") {
				return step["content"] ?? "";
			}
		}
		return null;
	}

	getCurrentTurnIdx(): number {
		for (const step of this.streamStepsReverse(1000)) {
			if (step["type"] === "USER_INPUT") {
				return step["step_index"] ?? 0;
			}
		}
		return 0;
	}

	getUserInputCount(): number {
		let count = 0;
		for (const step of this.streamStepsForward()) {
			if (step["type"] === "USER_INPUT") {
				count += 1;
			}
		}
		return count;
	}

	hasPb(): boolean {
		return fs.existsSync(this.pbPath);
	}

	getPbRoot(): Record<number, any[]> | null {
		if (!fs.existsSync(this.pbPath)) return null;
		try {
			const encrypted = fs.readFileSync(this.pbPath);
			const decrypted = decryptPb(encrypted);
			return parseProtobuf(decrypted);
		} catch {
			return null;
		}
	}

	getPbStepCount(): number | null {
		const root = this.getPbRoot();
		if (!root) return null;
		return (root[4]?.[0] as number) ?? null;
	}

	getLastModifiedTime(): number {
		let t = 0;
		if (fs.existsSync(this.dbPath)) {
			try {
				t = fs.statSync(this.dbPath).mtimeMs / 1000;
			} catch {
				// pass
			}
		}
		if (fs.existsSync(this.pbPath)) {
			try {
				const pt = fs.statSync(this.pbPath).mtimeMs / 1000;
				if (pt > t) t = pt;
			} catch {
				// pass
			}
		}
		return t;
	}
}
